File: java/tutorials/tour-of-restate-java/src/main/java/dev/restate/tour/part4/CheckoutService.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Java,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

package dev.restate.tour.part4;

import dev.restate.sdk.JsonSerdes;
import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.tour.auxiliary.CheckoutRequest;
import dev.restate.tour.auxiliary.EmailClient;
import dev.restate.tour.auxiliary.PaymentClient;

@Service
public class CheckoutService {
    // <start_checkout>
    @Handler
    public boolean handle(Context ctx, CheckoutRequest request) {
        double totalPrice = request.getTickets().size() * 40.0;

        String idempotencyKey = ctx.random().nextUUID().toString();
        boolean success = ctx.run(JsonSerdes.BOOLEAN, () ->
                PaymentClient.get().call(idempotencyKey, totalPrice));

        if (success) {
            ctx.run(()-> EmailClient.get().notifyUserOfPaymentSuccess(request.getUserId()));
        } else {
            ctx.run(() -> EmailClient.get().notifyUserOfPaymentFailure(request.getUserId()));
        }

        return success;
    }
    // <end_checkout>
}

File: typescript/tutorials/tour-of-restate-typescript/src/part4/checkout_service.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

import * as restate from "@restatedev/restate-sdk";
import { PaymentClient } from "../auxiliary/payment_client";
import { EmailClient } from "../auxiliary/email_client";

export const checkoutService = restate.service({
  name: "CheckoutService",
  handlers: {
    // <start_checkout>
    async handle(ctx: restate.Context, request: { userId: string; tickets: string[] }) {
      const totalPrice = request.tickets.length * 40;

      const idempotencyKey = ctx.rand.uuidv4();
      const success = await ctx.run(() => PaymentClient.get().call(idempotencyKey, totalPrice));

      if (success) {
        await ctx.run(() => EmailClient.get().notifyUserOfPaymentSuccess(request.userId));
      } else {
        await ctx.run(() => EmailClient.get().notifyUserOfPaymentFailure(request.userId));
      }

      return success;
    },
    // <end_checkout>
  }
});

export const CheckoutService: typeof checkoutService = { name: "CheckoutService"};
